libraries/chain/include/eosio/chain/snapshot.hpp

```diff
@@ -406,4 +411,16 @@ namespace eosio { namespace chain {
 
    };
 
+   struct snapshot_loaded_row_counter {
+      snapshot_loaded_row_counter(const size_t total) : total(total) {}
+      void progress() {
+         if(++count % 50000 == 0 && time(NULL) - last_print >= 5) {
+            ilog("Snapshot initialization ${pct}% complete", ("pct",(unsigned)(((double)count/total)*100)));
+            last_print = time(NULL);
+         }
+      }
+      size_t count = 0;
+      const size_t total = 0;
+      time_t last_print = time(NULL);
+   };
```

Comment:
It's a little tempting to move this completely inside of snapshot_reader so that maybe the onus to call `progress()` remains entirely inside of the `snapshot_reaader` & friends (which would probably help ensuring there isn't missed calls). But `snapshot_reader`'s interface technically allows you to read the same section multiple times (even though nobody does that..) so, eh.. not a perfect match.

Reply:
I could see a desire for a user to provide their own version of `snapshot_loaded_row_counter` say for diff logging purposes. Keeping it external would make that easier. I don't think it worth making it a template or derivable type at this point until that is needed.

Reply:
Good idea buuuutt.. this all probably needs to be refactored again when making loading parallel so that might not be possible. I didn't mention that this may be somewhat temporary since I'm not committing to parallel loading just yet..